### PR TITLE
Stress Framework

### PR DESCRIPTION
**Pull request by mikeharder**
Framework: `common/Stress/Azure.Test.Stress`
Sample Tests used when developing framework: `common/Stress/Azure.Sample.Stress`
Template Tests used as template for per-service stress tests: `sdk/template/Azure.Template/stress`

I would like to get this merged by EOD tomorrow at the latest, so developers can start writing tests next week.  If there is feedback on the csproj files, directory names, etc I am happy to make those changes now.  If there is feedback on the details of the framework I would prefer to make these changes in a later PR.

### CODE REVIEW

File: common/Stress/Azure.Test.Stress/Azure.Test.Stress.csproj
Comment (pakrym) [inline common/Stress/Azure.Test.Stress/Azure.Test.Stress.csproj:4]: nit: trailing ws

File: sdk/template/Azure.Template/stress/Directory.Build.props
Comment (pakrym) [inline sdk/template/Azure.Template/stress/Directory.Build.props:3]: We shouldn't have to do this for actual stress projects. Can we add a convention based on the project name?

File: sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj
Comment (pakrym) [inline sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj:1]: I wonder if it's too early to add this project to the template. The template is used by onboarding teams to start a library and I doubt any of them needs a stress project just yet.

**Review by pakrym — APPROVED** [review]

**Comment by pakrym** [comment]
Are tests for the stress framework planned?

**Comment by pakrym** [comment]
I think we should run an iteration of every stress test as part of live tests runs to make sure they are functioning.

File: sdk/template/Azure.Template/stress/Directory.Build.props
Comment (weshaggard) [inline sdk/template/Azure.Template/stress/Directory.Build.props:2]: I wouldn't mark this as a test support project. instead I would just rename the project to be Azure.Template.Stress.Tests.csproj and that will cause it to be treated as a test project. If we need special handling for these then we can add a new IsStressTestProject and default it based on the filename "Stress.Tests".

File: sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj
Comment (weshaggard) [inline sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj:1]: For the directory structure should we put it under "tests"? I guess if we do there will be some work to keep the projects independent because most tests put their project in the root of the tests folder. I'm Ok with using "<package>/stress", just wanted to mention it.

File: eng/Packages.Data.props
Comment (weshaggard) [inline eng/Packages.Data.props:191]: This sets the version but I didn't see you actually reference these package in your stress project.

File: sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj
Comment (mikeharder) [reply sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj:1]: I do want a template showing how to add stress tests though.  Where else could it live?  And the onboarding training does include perf and stress, so I think it's good for teams to at least think about right away.

**Comment by mikeharder** [comment]
@pakrym:

> Are tests for the stress framework planned?

I wasn't planning on adding them.  Do you think they are required?  This would be a new requirement for all languages.

**Comment by pakrym** [comment]
> Do you think they are required?

I do. How would we know it's working when making changes to the framework? I don't believe in manual testing.

**Comment by mikeharder** [comment]
@pakrym:

> I think we should run an iteration of every stress test as part of live tests runs to make sure they are functioning.

Once the perf/stress automation is running, this will run the perf/stress tests nightly to both collect results and ensure they are functioning.  I think adding this to the live test pipeline as well would be overkill, at least for now.

**Comment by mikeharder** [comment]
> > Do you think they are required?
> 
> I do. How would we know it's working when making changes to the framework? I don't believe in manual testing.

I agree in principle, but this is a fairly large work item across all languages and will need to be prioritized against things like the perf/stress automation.  We will need to add the frameworks to the repo without self-tests to unblock the service tests.

File: sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj
Comment (mikeharder) [reply sdk/template/Azure.Template/stress/Azure.Template.Stress.csproj:1]: I'm open to whatever you or the language team wants here.  I made "stress" a sibling of "tests" for exactly this reason, I didn't want to break any existing tests.

File: eng/Packages.Data.props
Comment (mikeharder) [reply eng/Packages.Data.props:191]: They are only needed by the stress framework.

File: sdk/template/Azure.Template/stress/Directory.Build.props
Comment (mikeharder) [reply sdk/template/Azure.Template/stress/Directory.Build.props:2]: How about I add `IsStressProject` (and later `IsPerfProject`), and condition this based on the name `*.Stress`?  The perf and stress tests are fairly different than the unit and functional tests, and I think it will be cleaner to consider them a sibling rather than nesting under "tests".  Similar to how "samples" and "tests" are siblings.

File: sdk/template/Azure.Template/stress/Directory.Build.props
Comment (mikeharder) [reply sdk/template/Azure.Template/stress/Directory.Build.props]: Added `IsStressProject`, set by naming convention.

File: sdk/template/Azure.Template/stress/Directory.Build.props
Comment (mikeharder) [reply sdk/template/Azure.Template/stress/Directory.Build.props]: Added `IsStressProject`, set by naming convention.

**Comment by AlexGhiondea** [comment]
@mikeharder I think we should have tests for this. I understand that the work to add tests needs to be prioritized against other things, but if that is the case we should have a work item tracking that and make sure it gets triaged accordingly (i.e. we will get to it sooner rather than later 🙂 )

**Comment by mikeharder** [comment]
> @mikeharder I think we should have tests for this. I understand that the work to add tests needs to be prioritized against other things, but if that is the case we should have a work item tracking that and make sure it gets triaged accordingly (i.e. we will get to it sooner rather than later 🙂 )

Add tests for Perf and Stress framework #16697

I will add tests for the .NET frameworks.  Tests for the other languages will be delegated to the per-language perf/stress maintainers.

**Comment by mikeharder** [comment]
@weshaggard, @pakrym: Are you OK with me merging this today?

File: eng/Directory.Build.Data.props
Comment (weshaggard) [inline eng/Directory.Build.Data.props:42]: I'm good with this convention but it looks like it will be set for both the the stress framework library as well as the stress tests projects. There might be some cases where you need to distinguish between the 2.

**Review by weshaggard — APPROVED** [review]
Overall looks reasonable to me.

File: eng/Directory.Build.Data.props
Comment (mikeharder) [reply eng/Directory.Build.Data.props:42]: This was intentional since their settings should be the same, I can add `IsStressSupportProject` later if needed.  I'd prefer to keep perf and stress separate from "test" since their settings are different.

**Pull request merged**
